data-raw/simulate-data.R

```diff
@@ -0,0 +1,243 @@
+  #       "A10BX"
+  # )
+  data |>
+    mutate(
+      across(
+        matches("^atc$"),
+        \(column) if_else(
+          runif(n()) < proportion,
+          sample(c("A10BA02", "A10BJ02", "A10BJ06"), 1),
+          column
+        )
+      )
+    )
+}
+
+# Insert a few false-positive cases with purchases of metformin:
+insert_false_metformin <- function(data, proportion = 0.05) {
+  if (!all(colnames(data) %in% c("atc", "name"))) {
+    return(data)
+  }
+  data |>
+    dplyr::mutate(
+      atc = dplyr::if_else(
+        indo %in% c("0000092", "0000276", "0000781") & insertion_rate(proportion),
+        "A10BA02",
+        atc
+      ),
+      name = dplyr::if_else(
+        indo %in% c("0000092", "0000276", "0000781") & insertion_rate(proportion),
+        "metformin",
+        name
+      )
+    )
+}
+
+# Insert some false positives for Wegovy and Saxenda.
+insert_false_drug_names <- function(data, proportion = 0.05) {
+  if (!all(colnames(data) %in% c("atc", "name"))) {
+    return(data)
+  }
+  data |>
+    mutate(
+      name = case_when(
+        atc == "A10BJ06" & insertion_rate(proportion) ~ "Wegovy Flextouch",
+        atc == "A10BJ02" & insertion_rate(proportion) ~ "Saxenda",
+        TRUE ~ name
+      )
+    )
+}
+
+insert_analysiscode <- function(x) {
+  # TODO: Is this necessary?
+  # ANALYSISCODE: npu code of analysis type (chr)
+  # 50% is either NPU27300 or NPU03835
+  # other 50% is NPU10000 to NPU99999
+  x
+}
+
+# Simulate data -----------------------------------------------------------
+
+# use the simulation definition data to simulate some data
+simulate_data <- function(data, n) {
+  # N needs to be capitalized for fabricatr, and but to be consistent
+  # with other functions and their use of `n`, I kept it lowercase for
+  # the function argument.
+  N <- n
+  data$generator |>
+    as.list() |>
+    set_names(data$variable_name) |>
+    # this evaluates the character string
+    map(~ eval(parse(text = .x))) |>
+    # enframe converts a list to a tibble
+    imap(\(column, name) enframe(column, name = NULL, value = name)) |>
+    unname() |>
+    list_cbind()
+}
+
+add_fake_drug_name <- function(data) {
+  data |>
+    mutate(
+      across(
+        matches("^name$"),
+        \(x) create_fake_drug_name(atc = atc)
+      )
+    )
+}
+
+set.seed(123)
+simulation_definitions <- here("data-raw/simulation-definitions.csv") |>
+  read_csv(show_col_types = FALSE) |>
+  select(register_abbrev, variable_name, generator)
+
+simulation_definitions_list <- simulation_definitions |>
+  group_split(register_abbrev)
+
+register_data <- simulation_definitions_list |>
+  map(\(data) simulate_data(data, n = 1000)) |>
+  map(insert_specific_atc) |>
+  map(add_fake_drug_name) |>
+  map(insert_false_drug_names) |>
+  map(insert_false_metformin) |>
+  # add the register abbreviations as a name to the list
+  set_names(
+    map_chr(
+      simulation_definitions_list,
+      \(data) unique(data$register_abbrev)
+    )
+  )
```

Comment:
Man, this is so sweet!

Reply:
A few details, just putting them here for future reference:

- sysi should cover years 1990 through 2005, and sssy 2005 and onward. It only matters to the preprocessing steps (handling the duplicated data during 2005 where the two tables overlap) but it makes no difference to the algorithm itseflf, so it's just a potential tweak to the last 2 digits of `honuge` if we want to more closely mimic the real data.
- The NPU-code values should be 8 digits long, including the "NPU"-prefix. Right now they're 11 digits in total length ("NPU"+8 digits).
- In the LPR3 data (the `diagnoser` table), the variable diagnosetype is currently 0 or 1, but should be "A" or "B", similar to the LPR2 data (the `lpr_diag` table).
- As we mentioned on the meeting last week, we probably need to reuse the same vector of `recnum`s in both the `lpr_adm` + `lpr_diag` tables (and, similarly, reuse an `dw_ek_kontakt` vector for both LPR3 tables), similar to how the pnrs are the same in all the tables. Otherwise joining the tables using `recnum`/`dw_ek_kontakt`  as the joining key is going to be messy.
- The vnr values for medications probably have to be hardcoded to match the real data/edge cases. I'll look them up and get back to you.